### DISCUSSION TITLE
Открытие порта 554

### DISCUSSION THREAD

**Discussion by mgorshkov** [Arkady23/V380-Web#48] [General]
Добрый день!
Имею камеру V380.
Software version: AppV380E31_V1.1.6.3_202111105
Firmware version: HwV380E31_WF8_PTZ_WIFI_20201218
Задача: открыть порт 554 (RTSP).
Mikhails-MacBook-Pro:~ mgorshkov$ telnet 192.168.1.9 554
Trying 192.168.1.9...
telnet: connect to address 192.168.1.9: Connection refused
telnet: Unable to connect to remote host
Mikhails-MacBook-Pro:~ mgorshkov$ 
Списывался с официалами, задав вопрос, как открыть порт. Они сначала прислали ceshi.ini, он не помог, затем прошивку 
SPECIAL_V380_AK_3918E-V330_20MB-DMA_V9.9.9.9_2021-09-24.zip, она тоже не помогла.
Затем официалы замолчали.
Пробовал последний архив отсюда, он тоже не помог, насколько помню.
Есть возможность перехватить трафик с камеры. В нем есть строки
Mikhails-MacBook-Pro:~ mgorshkov$ cat out.txt | grep param
P. C.[....o...E.....@.@......Np|....".`$oP..-	P. .}...param=ewogICJ1c3IiIDogIllXUnRhVzQ9IiwKICAiZGlkIiA6IDU4MDE3Mzk4LAogICJjb3VudCIgOiAxMCwKICAia2V5IiA6ICJhIiwKICAidHlwZSIgOiAwLAogICJwd2QiIDogIjU5NTBiOGQ3MThmODczNGE3OTUxZjlhYWNmMzFjOTkwIiwKICAiZnRpbWUiIDogMTY1ODYxMDAwMDQyOAp9&type=1
Mikhails-MacBook-Pro:~ mgorshkov$ cat out.txt | grep V380
User-Agent: V380%20Pro/62 CFNetwork/1333.0.4 Darwin/21.5.0
......>...........k..AppV380E31_V1.1.6.3................................................................................................................................................................................................................................
Больше ничего, связанного с param или V380, не нахожу.
Снимаю трафик через 
sudo tcpdump -i rvi1 -A > out.txt
rvi1 - виртуальный интерфейс iPhone, на котором запущено приложение, где я нажимаю "Firmware version".
Посоветуйте, пожалуйста, как открыть порт 554.
Спасибо!

**Comment by Arkady23** [comment]
Я rtsp открываю через vlc. Вы можете прислать их прошивку SPECIAL_V380_AK_3918E-V330_20MB-DMA_V9.9.9.9_2021-09-24.zip? Вы уверены, что эта прошивка отработала? Её работа была видна, скажем на изменениях на sd, или были сообщения при обновлении...

**Comment by mgorshkov** [comment]
Добрый день! Да, прошивка 100% отработала. Были сообщения при обновлении
(software update starting, software update completed, system starting,
system started, system connected). Прилагаю прошивку аттачем. После
прошивки, как и до прошивки, порт не открылся.
Что с SD до и после, есть ли какая-то разница, я не проверял. Можно
проверить теоретически говоря.
[SPECIAL_V380_AK_3918E-V330_20MB-DMA_V9.9.9.9_2021-09-24.zip](https://github.com/Arkady23/V380-Web/files/9175825/SPECIAL_V380_AK_3918E-V330_20MB-DMA_V9.9.9.9_2021-09-24.zip)

**Reply by Arkady23** [reply]
Я изучил ваш официальный патч SPECIAL_V380_AK_3918E-V330_20MB-DMA_V9.9.9.9_2021-09-24.zip.
И на его логике сделал подобный патч с web-сервером. Установите прилагаемый патч (прошивку), у вас должен появится доступ к командам linux камеры через файл startup.sh (расположенный на sd в папке ark-add-on), который будет выполняться при включении камеры. В нем в качестве примера будет запускаться http-сервер. Протестируйте web-сервер, протестируйте telnet. Главное, чтобы этот патч отработал, остальное появится. Rtsp может включатся и выключается в настройках через web. Обязательно напишите, что и как произошло, как работает.....
[httpd_V380E4_C3_9090909_v20210620.zip](https://github.com/Arkady23/V380-Web/files/9176820/httpd_V380E4_C3_9090909_v20210620.zip)

**Reply by mgorshkov** [reply]
Спасибо, @Arkady23! Проверю, отпишусь по результатам.

**Reply by mgorshkov** [reply]
Отформатировал флешку (128Gb, exFat, ранее так же форматировал, и она работала), распаковал в нее приложенный архив (см аттач). Запустил камеру, она сказала: system starting, system startup completed, system connected, но ничего про software update. Проверил порты - ни один новый порт не открылся, 554 и 80 недоступны (через веб-интерфейс не открывается), порт телнета тоже.
Что посоветуете?
![patch](https://user-images.githubusercontent.com/5475799/180730264-9a7e4960-3164-4109-9035-7688c99db6bf.png)

**Reply by Arkady23** [reply]
Странно. Либо китайцы что-то изменили в формате патча, либо надо fat32. Насколько я знаю, только fat32 обеспечивает права 777.

**Reply by mgorshkov** [reply]
Могу попробовать и FAT32, но у меня только 4Gb флешка для FAT32, а хотелось бы использовать на 128, потому что туда же записывается и видео с камеры.
Китайский патч работал с exFat и флешкой на 128 (работал в плане того, что он аплаился, но не делал того, что мне нужно).
Предположу, что действительно китайцы что-то изменили, но попробую и FAT32. Отпишусь по результатам.

**Reply by Arkady23** [reply]
Главное, чтоб патч сработал, а дальше sd можно поменять.

**Reply by mgorshkov** [reply]
Не заработало. Я нашел флешку на 1Gb, отформатировал как FAT (по дефолту), распаковал туда архив, вставил в камеру. Камера похоже что опять не увидела патч (system starting, started, connected). Порты не открылись, порт 80 недоступен. Сама sd карта видна камерой, доступна через мобильное приложение.
А насчет "заменить sd" - правильно я понимаю, что ваш патч работает только "вживую" на sd, а если убрать sd, то он не активируется? То есть изменения в камеру не вносятся, как написано в описании к репо? Следовательно, с "чистым" sd я буду иметь и чистую камеру без патча.

**Reply by Arkady23** [reply]
Да, но что об этом говорить, если патч не активируется. На самом деле я использовал утилиту формирования патча, выложенную на 4pda. Патчи работали до определенного времени. Уже пару лет, как перестали работать. Возможно надо проанализировать изменился ли формат  патча. Мой патч записывает на саму камеру несколько строк типа:
```
if [ -f "/mnt/sdcard/ark-add-on/startup.sh" ]; then
  /mnt/sdcard/ark-add-on/startup.sh &
fi

```

**Reply by Arkady23** [reply]
Если файл startup.sh присутствует на sd, то он запускается в фоновом режиме. Если для sd нет доступа 777, то с sd ничего запускаться не будет, а как данные sd может использоваться и при этом не надо допуска 777. Допускаю, что для активации патча три семерки не обязательны.
Идет борьба хакеров с китайскими производителями ПО. Взломщики пытаются получить доступ, чтобы что-то свое записать в камеру. Как только китайцы узнают, что в их систему что-то могут записать, они применяют меры и что-то изменяют. Как вариант, они могли изменить правило подсчета контрольной суммы патча. А перед тем, как устанавливать патч утилита установки патча проверяет контрольную сумму данных патча.

**Reply by mgorshkov** [reply]
Может тогда просто взять за основу прошивку от официалов, и добавить к ней ваш загрузчик вместе с веб-сервером и тем что надо? Но у меня более насущный вопрос: как открыть порт? Я могу взять прошивку от официалов, в ней что-то изменить и получить открытый порт? Каким образом порты открываются в вашей прошивке? Через iptables/ipfw?

**Reply by Arkady23** [reply]
Порт открывает сервер RTSP. Он в камере. Web-сервер, только меняет параметры камеры. Стал известен пароль рут. Поэтому сейчас думаю изменить описание по установке для новых камер через телнет. Я так понимаю, что RTSP у вас должен запускаться.

**Reply by mgorshkov** [reply]
Спасибо! Жду тогда обновления описания.

**Comment by Arkady23** [comment]
Добавил описание для установки httpd вручную для новых камер.

**Reply by mgorshkov** [reply]
Спасибо! Попробую позже и отпишусь по результатам.

**Reply by mgorshkov** [reply]
При скачке по ссылке https://github.com/Arkady23/V380-Web/releases/download/20210530/httpd_V380_any_manual_v20210620.zip из README выводится "Not found".

**Reply by Arkady23** [reply]
У вас старая ссылка. Ссылки обновлены..... Понял. По тексту еще старые ссылки остались. Сейчас исправлю.....Исправил.

**Reply by mgorshkov** [reply]
Спасибо! Я скачал архив, распаковал на SD-карту, камера загрузилась, и телнет заработал! Ура!! Запустил стартап-скрипт. Камера перезагрузилась, и стала доступна через веб.
Теперь у меня есть веб-интерфейс, но он ничего не показывает и кажется не работает (я пробовал включать RTSP, нажимал Save Settings, потом делал reboot, но камера не ребутится и ничего не меняется. Телнет продолжает к счастью работать, можно им пользоваться).
В принципе, мне нужно только узнать как из телнета включить RTSP, либо открыть порт 554 для доступа снаружи, если RTSP сервер уже есть в камере и работает. Но от веб интерфейса тоже не откажусь!
Еще раз спасибо за помощь! Прикладываю скриншоты, как у меня выглядит веб-интерфейс.
![1](https://user-images.githubusercontent.com/5475799/183251649-8c838f53-4d63-4b63-a2dd-13346ecdb423.png)
![2](https://user-images.githubusercontent.com/5475799/183251658-5cec03a4-7d5b-493d-af7a-d86199193399.png)
![3](https://user-images.githubusercontent.com/5475799/183251772-266e190c-f3c7-47eb-826d-f6e0d8004fb5.png)

**Reply by Arkady23** [reply]
Можете скопировать весь линукс на sd-карту? Может с помощью tar...

**Reply by Arkady23** [reply]
Еще может быть интересна папка с видео. Не само видео, структура папок и имена файлов.

**Reply by Arkady23** [reply]
Еще про ftp не написали. ftp работает?

**Reply by Arkady23** [reply]
Включение rtsp зависит от настроек в файле /mnt/mtd/mvconf/factory_const.ini
Там должна быть в секции [CONST_PARAM] строка примерно такая rtsp=1, если включено, и rtsp=0, если rtsp выключен.

**Reply by Arkady23** [reply]
Если строки с rtsp в секции [CONST_PARAM] нет, то скорей всего в камере нет rtsp совсем.

**Reply by Arkady23** [reply]
У мня есть предположение, что в новых камерах другой Линукс. Хотелось бы получить из вашей ос библиотеки, которые я использую, возможно они у вас другие:
ld-uClibc.so.0
libc.so.0
libm.so.0

**Reply by Arkady23** [reply]
Еще может быть вариант, что в камере не хватает ресурсов. Вы выключите telnet, ftp, rtsp оставьте только один httpd. И проверьте еще раз его работу.

**Reply by mgorshkov** [reply]
Завтра посмотрю, и на все отвечу. Спасибо!

**Reply by mgorshkov** [reply]
Итак, еще раз, что я делал вчера и что мы имеем сейчас:
1. Вчера я распаковал на SD архив с 
ark-add-on
quick_check.ini
setup.sh
2. Камера успешно соединилась через телнет
3. Я запустил /mnt/sdcard/setup.sh через телнет-сессию
4. Камера перезагрузилась и стал доступен веб-интерфейс
5. Веб-интерфейс не работал, после этого я написал сюда, снял скриншоты с веб-интерфейса и пошел спать
6. Утром telnet-сессия отвалилась и телнет порт стал недоступен (наверно камера ребутилась ночью)
7. Веб интерфейс все еще работает
8. Я заребутил камеру сам через мобильное приложение, она работает, но ничего не изменилось. Телнет не появился
9. Я решил перезалить архив на SD и перезагрузить камеру. Предварительно скопировал то, что на ней лежало
10. Это не помогло, что, как мне кажется, очень странно. Телнет не появился, веб интерфейс работает как раньше (то есть сам открывается, но в нем ничего не происходит)
Нужно ли прислать то, что я скопировал на шаге 9? Там есть какой-то диагностический файл. Может поможет.

**Reply by mgorshkov** [reply]
Подозреваю, что второй раз тот же патч уже не ставится, наверное что-то на камеру записывает. Может подскажете что надо изменить, чтобы телнет вернулся? Тогда смогу зайти и заархивировать весь линукс с камеры.

**Reply by Arkady23** [reply]
Второй раз ставить нет смысла. Патч ставится 1 раз. Если дело в том, что бы что-то запустить, то есть страртовый файл startup.sh (на sd-карте). Первый символ решетки закомментирует строку. Web может в зависимости от настроек сам ставить решетку на той или иной строке. Раскомментируйте строку с ftp, проверьте работу ftp. Сообщите. Все остальное пока закомментируйте. Строка с телнетом по умолчанию закомментирована.

**Reply by mgorshkov** [reply]
Да, спасибо, получилось! FTP работает, я зашел под рутом и паролем из файла quick_check.ini. Вижу директории
`bin
data
dev`
и т.д.
Могу скачать все к себе на комп, заархивировать и вам прислать. Структуру видео тоже могу прислать.
И если можно, как-то бы нам минимизировать количество вставок флешки и рестартов (камера находится на чердаке и прикручена снаружи, и каждый раз когда надо вставить флешку, я туда лажу по лестнице и с зеркалом вынимаю/вставляю флешку с риском уронить ее с чердака в огород).

**Reply by Arkady23** [reply]
Ну так удобней всего через ftp. Через ftp можно исправлять startup.sh. Пришлите Linux.

**Reply by Arkady23** [reply]
Наверное лучше через tar на sd Линукс запаковать, а потом чрез fp перписать.

**Reply by Arkady23** [reply]
Подозреваю, что всё-таки что-то с Линуксом. Давайте попробуем ваши библиотеки подсунуть. Скопируйте на вашу sd карту 3 библиотеки из папки /lib/ в папку sdcard/bin/, предварительно старые три файла с sd удалите
- ld-uClibc.so.0
- libc.so.0
- libm.so.0
Первую библиотеку ld-uClibc.so.0 переименуйте в просто lib

**Reply by Arkady23** [reply]
Включите web-сервер httpd и перезагрузитесь. Проверьте будет ли работать web-сервер.

**Reply by Arkady23** [reply]
Я копировал через ftp из lib на свой ПК, переименовал на ПК ld-uClibc.so.0, удалил через ftp эти 3 библиотеки и сразу скопировал с ПК их назад на sdcard/bin. У меня httpd включен..... я через него через меню перезагрузил камеру. С родными библиотеками у меня тоже все работает (и с теми, что в инсталяторе и с теми, что в системе).

**Reply by Arkady23** [reply]
На всякий случай попробуйте мои библиотеки из моего линукса.
[bin.zip](https://github.com/Arkady23/V380-Web/files/9276716/bin.zip)

**Reply by mgorshkov** [reply]
@Arkady23 посмотрите пока архив файлов с камеры (к сожалению, переписывал через FTP на Windows, затем запаковывал, но все же): [https://disk.yandex.ru/d/4lYAbT9x1ncJAQ](url)

**Reply by Arkady23** [reply]
Я тоже так делаю.

**Reply by Arkady23** [reply]
Скачал

**Reply by Arkady23** [reply]
Да, у вас Линукс более новый. И есть эти библиотеки новой версии..... Попробую их у себя потестировать.

**Reply by Arkady23** [reply]
За Линукс Спасибо!
Я у вас даже не нашел вот эту библиотеку: libm.so.0
Всё-таки попробуйте заменить библиотеки на библиотеки из моего линукса.

**Reply by mgorshkov** [reply]
libm.so.0 действительно у меня нету. Насчет заменить so - опасно, может не завестись.
Мне сейчас нужно просто включить RTSP. В factory_const.ini есть ключ rtsp_enable=10. Это наверное какие-то битовые флаги? Возможно, надо заменить на 11, чтобы включить? Как ваш веб-интерфейс включает RTSP?

**Reply by Arkady23** [reply]
Заменить не в системе, а на sd. Никакой опасности. Всё безопасно. На старых камерах всё заводится. Я вроде писал выше, что меняю rtsp=1 rtsp=0.

**Reply by mgorshkov** [reply]
Итак, заменил файлы на SD файлами из архива bin.zip. Результат отсутствует, ничего не изменилось. Веб-интерфейс показывает пустые страницы с настройками. Кнопка "сохранить" не работает.

**Reply by Arkady23** [reply]
Тут из соседней ветки дискуссии ответили, что web заработал. https://github.com/Arkady23/V380-Web/discussions/41#discussioncomment-3356772
Я ему дал немного другой бинарник web-сервера. Попробуйте. Может я его не так понял. И там переспросил на всякий случай.

**Reply by Arkady23** [reply]
Вот ссылка на еще один вариант httpd https://github.com/Arkady23/V380-Web/files/9206214/httpd2.zip

**Reply by mgorshkov** [reply]
[httpd2.zip](https://github.com/Arkady23/V380-Web/files/9206214/httpd2.zip) не заработал (камера по ip адресу стала недоступна через веб). Какой другой вариант - я не понял. Все бинарники кажется одинаковые:
$ ls -l httpd*
-rwxrwxr-x@ 1 mgorshkov  staff  294060 Jan  1  1980 httpd
-rw-rw-r--@ 1 mgorshkov  staff  294060 Jan  1  1980 httpd 2
-rw-rw-r--@ 1 mgorshkov  staff  294060 Jan  1  1980 httpd 3

**Reply by Arkady23** [reply]
Тут один тестер на 4pda описал ситуацию, что у него скрипты не работают. Т.е. по идее httpd запускается и работает, но не так, как надо. Оказалось он просто не обновил кеш обозревателя Ctrl+F5. Кроме того у него сбойная sd-карта. web-сервер из дистрибутива должен работать. Но всё-таки ошибку нашли — вид плейлиста не сохраняется. Скоро исправлю.

**Reply by mgorshkov** [reply]
Спасибо за комментарий, @Arkady23! Но у меня веб так нормально и не заработал. Однако на данный момент я решил свою проблему - открыл порт 554. Рискнул прописать в factory_const.ini rtsp_enable=1, и RTSP заработал.

**Reply by Arkady23** [reply]
Спасибо за информацию. Я тоже интуитивно смотрел на ваш rtsp_enable=0. Подумывал про него. Но камера у вас видимо необычная. Или может какая-то новая прошивка.

**Reply by Arkady23** [reply]
У вас только rtsp через web не включается? Возможно я могу в скрипте проанализировать камеру на параметр rtsp_enable

**Reply by Arkady23** [reply]
Думаю переделать скрипт, который меняет rtsp, чтобы он менял любую строку, где встречается rtsp в первой секции в квадратных скобках. Чтобы скрипт менял значение параметра не только для rtsp, но и для каrjго-нибудь rtsp1, rtsp1, rtsp_en, rtsp_enable или любой другой параметр где встречается rtsp. Скоро выложу обновление. Там же сделаю регулировку часового пояса.

**Reply by mgorshkov** [reply]
> Спасибо за информацию. Я тоже интуитивно смотрел на ваш rtsp_enable=0. Подумывал про него. Но камера у вас видимо необычная. Или может какая-то новая прошивка.

Было даже так:
`rtsp_enable=10`, я поменял на `rtsp_enable=1`, rtsp заработал. Видеопоток пошел (правда, камеру пришлось переключить с h265 на h264, иначе не работал RTMP), но аудио нет. Думаю, что это не принципиально.
Прошивка, вероятно, действительно новая. А камера обычная, купленная на озоне.

**Reply by mgorshkov** [reply]
> У вас только rtsp через web не включается? Возможно я могу в скрипте проанализировать камеру на параметр rtsp_enable

Ничего через веб не включается, интерфейс полностью мертвый, причем сам интерфейс показывается, но переключатели и кнопки не работают.

**Reply by mgorshkov** [reply]
> Думаю переделать скрипт, который меняет rtsp, чтобы он менял любую строку, где встречается rtsp в первой секции в квадратных скобках. Чтобы скрипт менял значение параметра не только для rtsp, но и для каrjго-нибудь rtsp1, rtsp1, rtsp_en, rtsp_enable или любой другой параметр где встречается rtsp. Скоро выложу обновление. Там же сделаю регулировку часового пояса.

Ок, спасибо, посмотрю.

**Comment by Arkady23** [comment]
Добрый день! Спасибо за Линукс. Постепенно что-то становится понятно. Дело, как я понял, в баше (sh), поэтому скрипты не работают. Для таких камер, как ваша попробовал sh поместить на sd карту. В принципе и старые камеры так тоже будут работать. Сделал новый вариант web-сервера, у меня он работает. Надо проверить на вашей камере. Прилагаю zip. просто скопируйте папки/файлы из zip на sd карту. Можно через ftp. Что бы точно знать, что у вас нет вирусов, проверьте длину файлов, хотя бы одного в папке bin, что бы быть уверенным, что скопировалось без искажений. Перезагрузитесь. Напишите, как web... заработал?
[web-server.zip](https://github.com/Arkady23/V380-Web/files/9637474/web-server.zip)

**Reply by mgorshkov** [reply]
Добрый день! Скопировал обновление через ftp, перезапустил камеру, веб показывает такую картинку:
<img width="1680" alt="Screenshot 2022-09-24 at 17 45 05" src="https://user-images.githubusercontent.com/5475799/192104287-a3402410-44d8-4891-b6ea-473269ed940b.png">

**Reply by mgorshkov** [reply]
Насколько вижу, дебаггер выдает ошибку 404: не найден файл http://192.168.1.9/cgi-bin/sdcard.sh

**Reply by Arkady23** [reply]
Почему тогда на старых камерах этот файл находится? Может что-то с правами? У вас же fat32? По-моему на ваших камерах busybox тоже не находится. Я завтра попробую еще 1 вариант web-сервера. Очень хорошо, что вы скриншот дали, хоть всё понятно с ситуацией. И кстати файл js загружается этим же web, по этому же пути.

**Reply by Arkady23** [reply]
Сейчас уже у многих такие камеры, как у вас.

**Reply by mgorshkov** [reply]
Не знаю почему на старых камерах этот файл находится. Сам файл есть на файловой системе: 
<img width="936" alt="Screenshot 2022-09-24 at 18 14 35" src="https://user-images.githubusercontent.com/5475799/192105704-aae4ff37-adaa-457a-8075-9c0605cbbdbe.png">
Раньше веб-интерфейс хотя бы показывался, но не работал (см старые скриншоты), а теперь не показывается.
У меня extFat (sdcard на 128Гб).

**Reply by Arkady23** [reply]
Попробуйте этот вариант. Я заменил в первом (чуть выше) варианте только папку bin
[web-server2.zip](https://github.com/Arkady23/V380-Web/files/9639078/web-server2.zip)

**Comment by mgorshkov** [comment]
Пока не заработало, наблюдаю то же самое. И ошибка та же самая:
 
<img width="1680" alt="Screenshot 2022-09-25 at 11 24 56" src="https://user-images.githubusercontent.com/5475799/192134727-5df8e6dc-6f06-431b-be60-5b51d54fe575.png">

**Reply by Arkady23** [reply]
Значит я н в ту сторону пошел. Пока всё возвращаю назад. Но не могу избавится от ощущения, что что-то с правами. Не нравится мне exFat. У меня эта фс не надежно работала на камере. Если будет что-то новое, я напишу.

**Comment by mgorshkov** [comment]
Ок, понял, попробую вечером.
Насчёт изменения параметра rtsp я бы предпочёл сначала узнать почему у меня
в данный момент там прописано 10, прежде чем менять.
